### ISSUE TITLE
[Bug]: Marlin W4A8 (int8 activations) corrupts on checkpoints with negative group scales

### ISSUE THREAD

**Issue by noonghunna** [vllm-project/vllm#48905]
### Your current environment

<details>
<summary>vLLM v0.25.1, RTX 3090 (Ampere, sm_86)</summary>

Marlin W4A8 (int8-activation) path via `VLLM_MARLIN_INPUT_DTYPE=int8` on a W4A16 checkpoint.
</details>

### 🐛 Describe the bug

With `VLLM_MARLIN_INPUT_DTYPE=int8`, a W4A16 checkpoint whose **weight group scales contain negative values** produces corrupt output (silent — no error, just garbage generations). AutoRound symmetric-int4 checkpoints legitimately emit ~50% negative group scales; these are perfectly valid for W4A16 (fp16 scales), but the int8-activation path corrupts them.

GPTQ / AWQ / llm-compressor recipes emit positive-only scales, which is why this hasn't surfaced.

**Root cause.** `marlin_act_int8_process_scales` (`marlin_utils.py`) packs the per-group scales into int16:

```python
def marlin_act_int8_process_scales(s: torch.Tensor):
    a_scales_scale_factor = 1 / 4096 * s.max().float()
    s = s / s.max() * 4096
    s = s.round().to(torch.int16).view(s.dtype)   # signed int16, negatives preserved
    return s, a_scales_scale_factor
```

so a negative scale is stored as a negative int16 (e.g. `-100 -> 0xFF9C`). But the Marlin kernel's int8-activation scale-application path reads these packed scales as **unsigned** — so `0xFF9C` is interpreted as `65436`, exploding the dequant by ~4 orders of magnitude for every group with a negative scale.

### Reproduce

Self-contained (no checkpoint needed — constructs symmetric int4 weights and toggles the scale sign):

```python
import torch
from vllm import _custom_ops as ops
from vllm.model_executor.layers.quantization.utils.marlin_utils import (
    apply_gptq_marlin_linear, marlin_act_int8_process_scales,
    marlin_make_workspace_new, marlin_permute_scales,
)
from vllm.scalar_type import scalar_types

torch.manual_seed(0); dev = "cuda"; k, n, g, m = 1024, 512, 128, 8
w_q = torch.randint(0, 16, (k, n), dtype=torch.int32, device=dev)
w_pack = torch.zeros(k // 8, n, dtype=torch.int32, device=dev)
for i in range(8): w_pack |= (w_q[i::8] & 0xF) << (4 * i)
empty = torch.empty(0, dtype=torch.int, device=dev); ws = marlin_make_workspace_new(torch.device(dev))
x = torch.randn(m, k, device=dev, dtype=torch.float16) * 0.5

def run(neg_fraction):
    scales = torch.rand(k // g, n, device=dev, dtype=torch.float16) * 0.004 + 0.002
    signs = torch.where(torch.rand_like(scales.float()) < neg_fraction, -1.0, 1.0).half()
    scales = scales * signs
    ref = x.float() @ ((w_q.float() - 8.0) * scales.float().repeat_interleave(g, 0))
    wq = ops.gptq_marlin_repack(w_pack, perm=empty, size_k=k, size_n=n, num_bits=4, is_a_8bit=True)
    s = marlin_permute_scales(scales.clone(), size_k=k, size_n=n, group_size=g, is_a_8bit=True)
    s, igs = marlin_act_int8_process_scales(s)
    out = apply_gptq_marlin_linear(input=x, weight=wq, weight_scale=s, weight_zp=empty, g_idx=empty,
        g_idx_sort_indices=empty, workspace=ws, wtype=scalar_types.uint4b8,
        output_size_per_partition=n, input_size_per_partition=k, is_k_full=True,
        input_global_scale=igs, bias=None, input_dtype=torch.int8)
    rel = ((out.float() - ref).abs() / ref.abs().clamp_min(1e-2)).median().item()
    print(f"{int(neg_fraction*100):3d}% negative group scales -> W4A8 median rel-err {rel:.4f}")

run(0.0)   # positive-only
run(0.5)   # 50% negative
```

Output on v0.25.1 / RTX 3090:

```
  0% negative group scales -> W4A8 median rel-err 0.0078
 50% negative group scales -> W4A8 median rel-err 16.3595
```

### Suggested fix

Read the int16-packed group scales as **signed** in the kernel's int8-activation scale-application path (the values written by `marlin_act_int8_process_scales` are signed int16). Alternatively, `marlin_act_int8_process_scales` could fold the sign into the weight codes (`q -> 16-q` for uint4b8) and store `|scale|`, but the signed read is the direct kernel-side fix.

Impact: any AutoRound-quantized W4A16 checkpoint is silently unusable with the int8-activation toggle.

**Comment by lazypool** [comment]
It sounds like this is a problem caused by the combination of AutoRound and the Marlin int8 activation kernel. Marlin assumes group scales are positive by default, which is also the general quantization assumption, but AutoRound violates this.

This seems like an easily fixable issue, and simply changing Marlin's uint16 to int16 would resolve it. It should be noted that doing so might increase the cost of dequantization, but considering the accuracy improvement that AutoRound brings, this tradeoff is still worthwhile. So I have submitted PR #48926 to address this. 

Thanks to @noonghunna for reporting this issue and for the suggested fix; I have added him as a co-author on the commit.

**Comment by kevinb361** [comment]
Additional real-model validation from a 2x RTX 3090 canary using a 27B AutoRound checkpoint with negative group scales.

The uncorrected INT8-activation path produced the expected corruption. After applying a sign-fold workaround (folding negative scale signs into the INT4 weight codes and storing positive scales), the model booted coherently and passed 30/30 structural and native-tool checks. There were no OOMs, parser failures, or unexpected restarts.

Cold direct-vLLM prefill improved by 1.614x at 11.45K input tokens, 1.488x at 55.24K, and 1.367x at 99.02K. That confirms the corrected path was actually engaged rather than silently falling back to W4A16.

My workaround is not the implementation in #48926, so this does not validate that PR's exact signed-read code. It does validate the issue on a full AutoRound model and shows that correcting the scale-sign handling restores coherent W4A8 operation with substantial Ampere prefill gains.

One operational note: the exact serving shape had a mean decode ratio of 0.892x versus W4A16 and speculative-token acceptance dropped from 65.55% to 61.57%. Those are separate performance tradeoffs; the scale correction itself fixed the corruption.
